### DISCUSSION TITLE
How to return prisma promise from function for Transaction

### DISCUSSION THREAD

**Discussion by RProduction** [prisma/prisma#21225] [Q&A] [answered]
### Question

Hello, I have some question.
How can I return prisma promise from db."repo"."method"?
I am using prisma with repository pattern, so I create my implementation inside invidual repository file.
I want to easily chain invidual function that perform crud inside transaction.
I don't want to rewrite crud method inside single transaction, so instead I want to be able to reuse all my repository function implementation.
Do someone know how to do this? probably like returning Prisma Promise or maybe passing transaction context to function so I can reuse all my implementation.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment, marked as answer]
Hi @RProduction 👋 

Thank you for raising this question.

Let's say you have these repository methods for your CRUD operations:
```js
const prisma = require('./db');

async function createUser(data) {
  return await prisma.user.create({
    data,
  });
}

async function updateUser(id, data) {
  return await prisma.user.update({
    where: { id },
    data,
  });
}
```
Now, you can use these repository methods within a [transaction](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#interactive-transactions). Here's an example of how to use transactions to chain multiple operations:
```js
async function performTransaction() {
  let result;

  try {
    result = await prisma.$transaction(async (tx) => {
      const user = await createUser({ name: 'John' });
      const updatedUser = await updateUser(user.id, { name: 'Jane' });

      // You can perform more operations within the transaction if needed

      return updatedUser;
    });
  } catch (error) {
    console.error('Transaction failed:', error);
    throw error;
  }

  console.log('Transaction completed:', result);
}

performTransaction()
  .catch((error) => {
    // Handle errors here
  })
  .finally(async () => {
    await prisma.$disconnect(); 
  });
```
In this example, `performTransaction` uses the `createUser` and `updateUser` repository methods within a single transaction. If any operation within the transaction fails, it will be rolled back.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by RProduction** [reply]
So you just need to wrap repository implementation inside that transaction? When I see ctx parameter in that lambda function, I thought I need to use that context.

Thank you very much for informing this.
Yeah, this discussion is resolved.
